keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -0,0 +1,1347 @@
+
+Currently, kubernetes Job API offers a way to account for infrastructure
+failures by setting `.backoffLimit > 0`. However, this mechanism intructs the
+job controller to restart all failed pods - regardless of the root cause
+of the failures. Thus, in some scenarios this leads to unnecessary
+restarts of many pods, resulting in a waste of time and computational
+resources. What makes the restarts more expensive is the fact that the
+failures may be encountered late in the execution time of a program.
+
+Sometimes it can be determined from containers exit codes
+that the root cause of a failure is in the executable and the
+job is destined to fail regardless of the number of retries. However, since
+the large workloads are often scheduled to run over night or over the
+weekend, there is no human assistance to terminate such a job early.
+
+The need for solving the problem has been emphasized by the kubernetes
+community in the issues, see: [#17244](https://github.com/kubernetes/kubernetes/issues/17244) and [#31147](https://github.com/kubernetes/kubernetes/issues/31147).
+
+Some third-party frameworks have implemented retry policies for their pods:
+- [TensorFlow Training (TFJob)](https://www.kubeflow.org/docs/components/training/tftraining/)
+- [Argo workflow](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml) ([example](https://github.com/argoproj/argo-workflows/blob/master/examples/retry-conditional.yaml))
+
+Additionally, some pod failures are not linked with the container execution,
+but rather with the internal kubernetes cluster management (see:
+[Scheduling, Preemption and Eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/pod-priority-preemption/)).
+Such pod failures should be recongnized as infrastructure failures and it
+should be possible to ignore them from the counter towards `backoffLimit`.
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+### Goals
+
+- Extension of Job API with user-friendly syntax to terminate jobs based on the
+  end state of the failed pod/container.
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+- Implementation of other indicators of non-retriable jobs such as termination logs.
+- Modification of the semantics for job termination. In particular, allowing for
+  all indexes of an indexed-job to execute when only one or a few indexes fail
+  [#109712](https://github.com/kubernetes/kubernetes/issues/109712).
+- Similar termination policies for other workload controllers such as Deployments
```

Comment:
I do think it's worth discussing whether this belongs at the job-spec level or the pod-spec level ... something like https://github.com/kubernetes/kubernetes/issues/31147 would require it at the pod-spec level